go/extra/ba/cmd/ba.go

```diff
@@ -0,0 +1,485 @@
+			maxThresholdAvgRatio: 1.3,
+			maxThresholdMaxRatio: 1.3,
+		},
+	}
+	userMetrics []string
+
+	logger = logging.GetLogger("cmd/ba")
+	client api.Client
+)
+
+type Metric struct {
+	getter               func(context.Context, string, *model.SampleStream) (float64, float64, error)
+	maxThresholdAvgRatio float64
+	maxThresholdMaxRatio float64
+	minThresholdAvgRatio float64
+	minThresholdMaxRatio float64
+}
+
+// getDuration returns average and maximum running times of the given coarse benchmark instance ("up" metric w/ minute
+// resolution time series).
+func getDuration(ctx context.Context, test string, bi *model.SampleStream) (float64, float64, error) {
+	instance := string(bi.Metric[testOasis.MetricsLabelInstance])
+
+	// Re-fetch the given benchmark instance with second resolution. Each obtained time series corresponds to one run.
+	v1api := v1.NewAPI(client)
+	r := v1.Range{
+		Start: bi.Values[0].Timestamp.Time().Add(-1 * time.Minute),
+		End:   bi.Values[len(bi.Values)-1].Timestamp.Time().Add(time.Minute),
+		Step:  time.Second,
+	}
+
+	query := fmt.Sprintf("up %s == 1.0", bi.Metric.String())
+	result, warnings, err := v1api.QueryRange(ctx, query, r)
+	if err != nil {
+		common.EarlyLogAndExit(fmt.Errorf("error querying Prometheus: %v", err))
+	}
+	if len(warnings) > 0 {
+		logger.Warn("warnings while querying Prometheus", "warnings", warnings)
+	}
+	if len(result.(model.Matrix)) == 0 {
+		return 0, 0, fmt.Errorf("getDuration: no time series matched test: %s and instance: %s", test, instance)
+	}
+	// Compute average and max duration of runs. Since we have a second-resolution, each point denotes 1 second of run's
+	// uptime. Just count all points and divide them by the number of runs.
+	avgDuration := 0.0
+	maxDuration := 0.0
+	for _, s := range result.(model.Matrix) {
+		avgDuration += float64(len(s.Values))
+		if maxDuration < float64(len(s.Values)) {
+			maxDuration = float64(len(s.Values))
+		}
+	}
+	avgDuration /= float64(len(result.(model.Matrix)))
+
+	return avgDuration, maxDuration, nil
+}
+
+// getIOWork returns average and maximum sum of read and written bytes by all workers of the given coarse benchmark
+// instance  ("up" metric).
+func getIOWork(ctx context.Context, test string, bi *model.SampleStream) (float64, float64, error) {
+	readAvg, readMax, err := getSummableMetric(ctx, "oasis_worker_disk_read_bytes", test, bi)
+	if err != nil {
+		return 0, 0, err
+	}
+	writtenAvg, writtenMax, err := getSummableMetric(ctx, "oasis_worker_disk_written_bytes", test, bi)
+	if err != nil {
+		return 0, 0, err
+	}
+
+	return readAvg + writtenAvg, readMax + writtenMax, nil
+}
+
+// getDiskUsage returns average and maximum sum of disk usage for all workers of the given coarse benchmark instance
+// ("up" metric).
+func getDiskUsage(ctx context.Context, test string, bi *model.SampleStream) (float64, float64, error) {
+	return getSummableMetric(ctx, "oasis_worker_disk_usage_bytes", test, bi)
```

Comment:
If metric names in `metrics.go` are moved into constants those could be reused here.